h2o-core/src/main/java/water/TCPReceiverThread.java

```diff
@@ -163,6 +163,10 @@ public void run() {
           // On any error from anybody, close everything
           System.err.println("IO error");
           Log.err("IO error on TCP port " + H2O.H2O_PORT + ": ", e);
+          if (e instanceof Error && H2O.isCI()) {
+            // could be AssertionError, OOM...
+            throw H2O.fail("Encountered an error while running in CI - will terminate to prevent deadlocks", e);
```

Comment:
is the idea to observe this on CI first, and then investigate+fix to avoid deadlocks in production?

Reply:
exactly, these kinds of issues frequently manifest themselves in our CI, and rarely we do see stuff like this reported from users

in CI I want to fail right away - because oftentimes people mistake these hangs for timeouts and they are not reported as issues

for non-CI (= production, use) - even if the cluster hangs on one specific algo - it doesn't make it generally unusable - some of these states are recoverable and we definitely don't want to shutdown the cluster that might have some completed work that wasn't saved